presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java

```diff
@@ -163,7 +171,7 @@ public boolean isPartitionedOn(Collection<Symbol> columns, Set<Symbol> knownCons
         // partitioned on (k_1, k_2, ..., k_n) => partitioned on (k_1, k_2, ..., k_n, k_n+1, ...)
         // can safely ignore all constant columns when comparing partition properties
         return arguments.stream()
```

Comment:
This is now incorrect. If partitioning is (a1, a2, expr1), then `isPartitionedOn([a1, a2])` should return false, since it's not a superset of the partitioning expressions.

Reply:
That would return false because `getColumn` on expr1 will return null which won't be in `columns`. But as I mentioned offline, that looks rather sketchy, since `isPartitionedOn([a1, a2, null)` will return true... 

I chatted with @haozhun offline and we think maybe it's just the logic is too complicated to reason about with `stream` so I rewrote it into a for loop. Hope that addressed your concern. 

My understanding of your initial comment is that you want to organize the logic into
1) Out of all arguments are expressions (including both `Symbol` and `Expression`), whether they are all column references (`Symbol`)
2) if yes, whether the given `columns` contains the symbol.

Please let me know whether this understanding is correct, or which part did I understand it wrong.

The problem with this logic is that there's no function to express "is either a symbol or an expression". There are a few ways to do that:
1) add another function to express this concept (which I did, but with wrong name)
2) `filter(argument -> argument.isVariable() || argument.isExpression())` or maybe `filter(argument -> !argument.isConstant())`, but neither gives the clear concept of "`variable` and `expression` should have both been expressions", which I think is what you want to express.
3) Change Symbol to SymbolReference everywhere so it's actually an Expression. This is a much bigger refactor, which doesn't have to be part of this PR. If we want to pursue this route, it should be a separate PR.

The for loop implementation is attempted at making it clear that there are 3 different kind of `ArgumentBinding` and we treat each separately. So at least there's no ambiguity in understanding, though it also doesn't address the idea of "both Symbol and Expression are expressions".

Please give clear instructions on how to proceed if you disagree with the current implementation. Thanks!

Reply:
> The problem with this logic is that there's no function to express "is either a symbol or an expression". 

Yes, that's what's bugging me a bit. My concern is that isPartitionedOn might start returning incorrect answers in the future once we generalize expressions/symbols, which would lead to either incorrect results or performance regressions. Hence why I was suggesting making isPartitionedOn insensitive to that by exposing to it the correct abstraction. I.e., everything looks to it like an expression to it, but some are symbol references and some are more complex expressions.

Reply:
Have you looked at what it would take to unify column & expression in ArgumentBinding? If it's not a lot of work, that might be the way to go.

Reply:
I didn't look into details, but I briefly checked what it takes. It would mean changing all public methods in `Partitioning`, `ActualProperties`, `PreferredProperties` that currently have `Symbol` as parameters, which seems like quite a lot. And it would also change the `translator` logic everywhere.

Reply:
The current implementation would somewhat address your concern about "returning incorrect answers in the future once we generalize expressions/symbols" because we are explicitly handling, and only handling the three cases as they currently are. So if we would change it to 2 cases in the future, that code would not work without revisiting this function. We can also add test cases to address your concerns if you have specific scenarios in mind. I've found the existing tests quite helpful (otherwise I wouldn't know I need to change the logic in this function).